finality-aleph/src/sync/handler.rs

```diff
@@ -0,0 +1,486 @@
+        &mut self,
+        justification: J,
+        peer: I,
+    ) -> Result<SyncActions<J>, Error<J, CS, V, F>> {
+        use JustificationAddResult::*;
+        let id = justification.header().id();
+        match self
+            .forest
+            .update_justification(justification, Some(peer))?
+        {
+            Noop => Ok(SyncActions::noop()),
+            Required => Ok(SyncActions::task(id)),
+            Finalizable => {
+                self.try_finalize()?;
+                Ok(SyncActions::noop())
+            }
+        }
+    }
+
+    /// Inform the handler that a block has been imported.
+    pub fn block_imported(&mut self, header: J::Header) -> Result<(), Error<J, CS, V, F>> {
+        match self.forest.update_body(&header)? {
+            true => self.try_finalize(),
+            false => Ok(()),
+        }
+    }
+
+    /// Handle a request for potentially substantial amounts of data.
+    ///
+    /// Currently ignores the requested id, it will only become important once we can request
+    /// blocks.
+    pub fn handle_request(
+        &mut self,
+        _requested_id: BlockIdFor<J>,
+        state: State<J>,
+    ) -> Result<SyncActions<J>, Error<J, CS, V, F>> {
+        let remote_top_id = self
+            .verifier
+            .verify(state.top_justification())
+            .map_err(Error::Verifier)?
+            .header()
+            .id();
+        Ok(SyncActions::response(
+            self.large_justification_batch_from(remote_top_id)?,
+        ))
+    }
+
+    /// Handle a single justification.
+    pub fn handle_justification(
+        &mut self,
+        justification: J::Unverified,
+        peer: I,
+    ) -> Result<SyncActions<J>, Error<J, CS, V, F>> {
+        let justification = self
+            .verifier
+            .verify(justification)
+            .map_err(Error::Verifier)?;
+        self.handle_verified_justification(justification, peer)
+    }
+
+    /// Handle a state broadcast returning the actions we should take in response.
+    pub fn handle_state(
+        &mut self,
+        state: State<J>,
+        peer: I,
+    ) -> Result<SyncActions<J>, Error<J, CS, V, F>> {
+        use Error::*;
+        let remote_top = self
+            .verifier
+            .verify(state.top_justification())
+            .map_err(Verifier)?;
+        let local_top = self.chain_status.top_finalized().map_err(ChainStatus)?;
+        match local_top
+            .header()
+            .id()
+            .number()
+            .checked_sub(remote_top.header().id().number())
+        {
+            // If we are just one behind then normal broadcasts should remedy the situation.
+            Some(0..=1) => Ok(SyncActions::noop()),
+            Some(2..=MAX_SMALL_JUSTIFICATION_BATCH) => Ok(SyncActions::response(
+                self.small_justification_batch_from(remote_top.header().id())?,
+            )),
+            Some(MAX_SMALL_JUSTIFICATION_BATCH_PLUS_ONE..) => Ok(SyncActions::response(
```

Comment:
hmm, seems like a good idea, don't remember why we decided to send only the top justification (maybe this was just to simplify the algorithm, not considering efficiency)

Reply:
We don't want to send a lot of justifications without an explicit request, sine they might be of nontrivial size.

Reply:
But we do that in case of a small justification batch. Why not do the same here with a small number of justifications?

Reply:
Yeah, but I understand that @maciejnems suggests sending the same set of justifications as in the `2..=MAX_SMALL_JUSTIFICATION_BATCH` case, plus one additional - this makes sense since now the peer expects at most `MAX_SMALL_JUSTIFICATION_BATCH` justifications in response to the state broadcast, so `+1` is not that much of a difference?
On the other hand, it's 2 messages faster, but may complicate the code, and the usual case is a node that's very far behind, so this will probably not make a difference? Hmm, the question is how often a node will lag slightly more than 10 justifications behind.

Reply:
Ah, I see the argument. IMO we should assume here that the difference is much larger than `MAX_SMALL_JUSTIFICATION_BATCH`, so we don't get much out of sending these justifications, while the code gets much more complex. If it turns out that we often are _just_ above this threshold then it would be much better to just increase the threshold.